Mathlib/RingTheory/TwoSidedIdeal/Operations.lean

```diff
@@ -0,0 +1,262 @@
+    s * t ⊆ I := by
+  rintro - ⟨x, hx, r, -, rfl⟩
+  exact mul_mem_right _ _ _ (h hx)
+
+lemma mem_span_iff_mem_addSubgroup_closure_nonunital {s : Set R} {z : R} :
+    z ∈ span s ↔ z ∈ AddSubgroup.closure (s ∪ s * univ ∪ univ * s ∪ univ * s * univ) := by
+  trans z ∈ span (s ∪ s * univ ∪ univ * s ∪ univ * s * univ)
+  · refine ⟨(span_mono (by simp only [Set.union_assoc, Set.subset_union_left]) ·), fun h ↦ ?_⟩
+    refine mem_span_iff.mp h (span s) ?_
+    simp only [union_subset_iff, union_assoc]
+    exact ⟨subset_span, subset_mul_set subset_span _, set_mul_subset subset_span _,
+      subset_mul_set (set_mul_subset subset_span _) _⟩
+  · refine mem_span_iff_mem_addSubgroup_closure_absorbing ?_ ?_
+    · rintro x y (((hy | ⟨y, hy, r, -, rfl⟩) | ⟨r, -, y, hy, rfl⟩) |
+        ⟨-, ⟨r', -, y, hy, rfl⟩, r, -, rfl⟩)
+      · exact .inl <| .inr <| ⟨x, mem_univ _, y, hy, rfl⟩
+      · exact .inr <| ⟨x * y, ⟨x, mem_univ _, y, hy, rfl⟩, r, mem_univ _, mul_assoc ..⟩
+      · exact .inl <| .inr <| ⟨x * r, mem_univ _, y, hy, mul_assoc ..⟩
+      · refine .inr <| ⟨x * r' * y, ⟨x * r', mem_univ _, y, hy, ?_⟩, ⟨r, mem_univ _, ?_⟩⟩
+        all_goals simp [mul_assoc]
+    · rintro y x (((hy | ⟨y, hy, r, -, rfl⟩) | ⟨r, -, y, hy, rfl⟩) |
+        ⟨-, ⟨r', -, y, hy, rfl⟩, r, -, rfl⟩)
+      · exact .inl <| .inl <| .inr ⟨y, hy, x, mem_univ _, rfl⟩
+      · exact .inl <| .inl <| .inr ⟨y, hy, r * x, mem_univ _, (mul_assoc ..).symm⟩
+      · exact .inr <| ⟨r * y, ⟨r, mem_univ _, y, hy, rfl⟩, x, mem_univ _, rfl⟩
+      · refine .inr <| ⟨r' * y, ⟨r', mem_univ _, y, hy, rfl⟩, r * x, mem_univ _, ?_⟩
+        simp [mul_assoc]
+
+end NonUnitalRing
+
+section Ring
+
+variable {R : Type*} [Ring R]
+
+open Pointwise Set in
+lemma mem_span_iff_mem_addSubgroup_closure {s : Set R} {z : R} :
+    z ∈ span s ↔ z ∈ AddSubgroup.closure (univ * s * univ) := by
+  trans z ∈ span (univ * s * univ)
+  · refine ⟨(span_mono (fun x hx ↦ ?_) ·), fun hz ↦ ?_⟩
+    · exact ⟨1 * x, ⟨1, mem_univ _, x, hx, rfl⟩, 1, mem_univ _, by simp⟩
+    · exact mem_span_iff.mp hz (span s) <| subset_mul_set (set_mul_subset subset_span _) _
+  · refine mem_span_iff_mem_addSubgroup_closure_absorbing ?_ ?_
+    · intro x y hy
+      rw [mul_assoc] at hy ⊢
+      obtain ⟨r, -, y, hy, rfl⟩ := hy
+      exact ⟨x * r, mem_univ _, y, hy, mul_assoc ..⟩
+    · rintro - x ⟨y, hy, r, -, rfl⟩
+      exact ⟨y, hy, r * x, mem_univ _, (mul_assoc ..).symm⟩
+
+/-- Given an ideal `I`, `span I` is the smallest two-sided ideal containing `I`. -/
+def fromIdeal : Ideal R →o TwoSidedIdeal R where
+  toFun I := span I
+  monotone' _ _ := span_mono
+
+lemma mem_fromIdeal {I : Ideal R} {x : R} :
+    x ∈ fromIdeal I ↔ x ∈ span I := by simp [fromIdeal]
+
+/-- Every two-sided ideal is also a left ideal. -/
+def asIdeal : TwoSidedIdeal R →o Ideal R where
+  toFun I :=
+  { carrier := I
+    add_mem' := I.add_mem
+    zero_mem' := I.zero_mem
+    smul_mem' := fun r x hx => I.mul_mem_left r x hx }
+  monotone' _ _ h _ h' := h h'
```

Comment:
Are `fromIdeal` and `asIdeal` a Galois connection? If so then you might want to write a TODO leaving this as future work (or proving it here, if you can come up with some three-line slick thing).

Reply:
Surely in that example I=R because it's generated by a unit?

Reply:
Indeed, let me try to prove the claim

Reply:
I have added the `gc`, sorry for the stupid example